Allow partial date parsing when simple datetime formatter is used

NEUpanning opened the pull request:
The Spark legacy datetime formatter allows parsing date from incomplete text, 
seeing [code link](https://github.com/openjdk/jdk8/blob/master/jdk/src/share/classes/java/text/DateFormat.java#L351). This PR enables partial date parsing when the `LENIENT_SIMPLE`
or `STRICT_SIMPLE` datetime formatter is used.

Relates issues: #10354, [gluten#6227](https://github.com/apache/incubator-gluten/issues/6227)

NEUpanning commented:
@rui-mo Could you please take a look? Thanks!

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -1588,8 +1588,9 @@ Expected<DateTimeResult> DateTimeFormatter::parse(
 
-  // Ensure all input was consumed.
-  if (cur < end) {
+  // Ensure all input was consumed if type_ is not legacy datetime formatter.
```
> legacy datetime formatter

Do we have a clear definition of which formatters are considered legacy?

NEUpanning replied:
Thanks. I updated the comment to use `LENIENT_SIMPLE or STRICT_SIMPLE datetime formatter` instead of `legacy datetime formatter`.

rui-mo commented on velox/functions/lib/tests/DateTimeFormatterTest.cpp:
```diff
@@ -2441,4 +2441,12 @@ TEST_F(SimpleDateTimeFormatterTest, formatWeekOfMonth) {
 }
 
+TEST_F(SimpleDateTimeFormatterTest, parseUsingPartialInput) {
+  for (bool lenient : {true, false}) {
```
nit: for easier debugging, we could write each test case separately.

NEUpanning replied:
Updated.

rui-mo reviewed (commented):
Thanks.

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -1588,8 +1588,10 @@ Expected<DateTimeResult> DateTimeFormatter::parse(
-  // Ensure all input was consumed.
-  if (cur < end) {
+  // Ensure all input was consumed if type_ is not LENIENT_SIMPLE or
+  // STRICT_SIMPLE datetime formatter.
```
Perhaps use `if type_ is not simple datetime formatter.` for simplicity.

NEUpanning replied:
Thanks. Updated.

rui-mo reviewed (approved):
Thanks.

NEUpanning commented:
@pedroerp Could you please take a look? Thanks!

NEUpanning commented:
@mbasmanova Could you please take a look? Thanks!

mbasmanova reviewed (commented):
@NEUpanning Thank you for the change. I assume this is a change in user-visible behavior. Would you update the docs?

NEUpanning commented:
@mbasmanova Thank you for your quick feedback. This change impacts the behavior of `unix_timestamp` and `get_timestamp` which relied on `DatetimeFormmater::parse`. I've added docs and rebased it onto the main branch.

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -117,7 +117,8 @@ These functions support TIMESTAMP and DATE input types.
     `Joda <https://www.joda.org/joda-time/>`_ date formatter depends on ``spark.legacy_date_formatter`` configuration.
     Returns NULL for parsing error or NULL input. When `Simple` date formatter is used, null is returned for invalid
-    ``dateFormat``; otherwise, exception is thrown. ::
+    ``dateFormat``; otherwise, exception is thrown. The `Simple` date formatter also permits partial date parsing;
```
> partial date parsing

Would readers know what this means? Would be nice to clarify a bit and provide an example.

NEUpanning replied:
@mbasmanova I've refined the documents and added function examples. I'm not sure if the examples meet the community standards. Could you please take a look?

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -312,8 +317,13 @@ These functions support TIMESTAMP and DATE input types.
     Returns null if ``string`` does not match ``format`` or if ``format``
-    is invalid. When `Simple` date formatter is used, permit partial date
-    parsing; otherwise, exception is thrown.
+    is invalid. The `Simple` date formatter permits partial date parsing
```
Thank you for clarifying the docs. I see that these docs repeat which creates a risk of them diverging. Would it be possible to remove copy-paste and refer to 'get_timestamp' function documentation. Alternatively, we may create a separate section that explains the 'format' and refer to it from all functions that use that 'format'.

NEUpanning replied:
@mbasmanova There are some different behaviors between `Simple` and `Joda` date formmater including how they handle invalid formats, whether the entire input needs to be consumed, and so on.  How about creating a new section to specifically clarify these differences?

mbasmanova replied:
> How about creating a new section to specifically clarify these differences?

That sounds good. Thanks.

NEUpanning replied:
@mbasmanova I have revised the documentation about the `Simple` date formatter for clarity and conciseness.  Could you please take a look?

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -337,3 +325,33 @@ These functions support TIMESTAMP and DATE input types.
         SELECT year_of_week('2005-01-02'); -- 2004
+
+Different Behaviors Between Simple And Joda Date Formmaters
+--------------------------------
```
I think --- needs to go all the way to cover the length of the title.

NEUpanning replied:
Thanks. Fixed

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -337,3 +325,33 @@ These functions support TIMESTAMP and DATE input types.
 
         SELECT year_of_week('2005-01-02'); -- 2004
+
+Different Behaviors Between Simple And Joda Date Formmaters
```
nit: Simple vs. Joda Date Formatter

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -337,3 +325,33 @@ These functions support TIMESTAMP and DATE input types.
+--------------------------------
+
+To align with Spark, Velox supports both `Simple <https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html>`_
+and `Joda <https://www.joda.org/joda-time/>`_ date formmater to parse/format timestamp/date strings
```
formmater -> formmaters

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -337,3 +325,33 @@ These functions support TIMESTAMP and DATE input types.
+
+To align with Spark, Velox supports both `Simple <https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html>`_
+and `Joda <https://www.joda.org/joda-time/>`_ date formmater to parse/format timestamp/date strings
+on the functions `from_unixtime`, `unix_timestamp`, `make_date` and `to_unix_timestamp`.
```
on the functions -> used in functions

for functions, use references 

```
:spark:func:`from_unixtime`
```

NEUpanning replied:
Thanks. Fixed

mbasmanova reviewed (approved):
@NEUpanning Thank you for iterating. Looks great.
